openedx/license-manager
# Add endpoint for sending license reminder emails

## Description

- Refactors how emails are sent
- Adds /licenses/remind endpoint which allows reminding a single user, specified by an email, that their license is pending activation
- Adds /licenses/remind-all endpoint which allows reminding all users of a subscription that have pending licenses that their licenses are pending activation

Still to come in separate PRs:
- Handle the emails in a celery task
- Handle errors in sending emails and only update `last_remind_date` on successful emails

Link to the associated ticket: https://openedx.atlassian.net/browse/ENT-2835

## Testing considerations
- Create a SubscriptionPlan with 3+ licenses at http://localhost:18170/admin/subscriptions/subscriptionplan/add/

### Test /remind validation
- Navigate to /api/v1/subscriptions/{subscription_uuid}/licenses/remind
- Verify that sending a POST request with no `user_email` in the data fails and an error is shown that says the email is required
- Verify that sending a POST request with `user_email` set to '' in the data fails and an error is shown that says the email is required / can not be blank
- Verify that sending a POST request with an invalid `user_email` ('lkajsdflkajsdf') in the data fails and an error is shown that says the email must be valid
- Verify that you can send a POST request with or without a custom 'greeting' or 'closing' in the data

### Test /remind functionality
- Assign a license to a user by navigating to http://localhost:18170/admin/subscriptions/license/ , changing a license associated with your subscription to have a status of assigned, and filling in the user email
- Navigate to /api/v1/subscriptions/{subscription_uuid}/licenses/remind
- Send a POST request with a custom greeting and closing, and the user email you assigned in the data.
  - Verify that the endpoint returns a 200
  - Verify that an email shows up in the logs that is sent to the specified user, has your custom greeting and closing, and shows a reminder message
  - Navigate to the license edit admin screen for the license you previously edited http://localhost:18170/admin/subscriptions/license/{license_uuid/change , and confirm that the `last_remind_date` field has been properly updated

### Test /remind-all functionality
- Assign a license to a different user in your same subscription plan through the same steps as before
- Change another license to have a status of `activated` and assign it to another user email
- Navigate to /api/v1/subscriptions/{subscription_uuid}/licenses/remind-all
- Send a POST request to the endpoint with a custom greeting and closing
  - Verify that the endpoint returns a 200
  - Verify that emails show up in the logs for all of the users that had pending licenses, but emails to the other users do not show up
  - Navigate to the admin screen for licenses and verify that all of the pending licenses in the subscription have had their `last_remind_date` fields updated accordingly, but the others have not
